fix: Correctly handle non str index list for bwa-mem2/mem

votti opened the pull request:
As reported in #522, currently the `bwa2-mem` wrapper has issues if `idx` is a sequence is it incorrectly checked `input.index` to be as string instead of `input.idx`. 

This  bug was silent, as usually `snakemake.input.index` would always be a available as `functools.partial(<function Namedlist._used_attribute at 0x73ac4dbc0c20>, _name='index')` or similar. Thus the predicate that it was not a `str` was always true.
Now with `snakemake=8.16.0` this changed and the bug becomes an error running the already existing tests.

While the discussion in #522 indicates, that there is an underlying issue with the concept of `input.idx` being as sequence of indices, this may need rework later. Until all the details are decided, having a correct working version should be still valuable. With the current state any workflow using this wrapper using `snakemake=8.16.0` is broken.

### QC
<!-- Make sure that you can tick the boxes below. -->

* [x] I confirm that I have followed the [documentation for contributing to `snakemake-wrappers`](https://snakemake-wrappers.readthedocs.io/en/stable/contributing.html).

While the contributions guidelines are more extensive, please particularly ensure that:
* [x] `test.py` was updated to call any added or updated example rules in a `Snakefile`
* [x] `input:` and `output:` file paths in the rules can be chosen arbitrarily
* [x] wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`)
* [x] temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to 
* [x] the `meta.yaml` contains a link to the documentation of the respective tool or command under `url:`
* [x] conda environments use a minimal amount of channels and packages, in recommended ordering


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **New Features**
  - Enhanced documentation clarifying input requirements for index files, ensuring users know all necessary file extensions.
  - Added a new function to extract the longest common prefix from index file names, improving error handling.

- **Bug Fixes**
  - Updated index file specifications to require a complete list of index files, preventing execution errors due to missing files.

- **Documentation**
  - Improved comments and descriptions in the Snakefile and meta.yaml for better user guidance on index file usage.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fgvieira commented:
I think that all input files should be specified as `input`, so that `snakemake` can check if they exist and are updated. Also, `idx` is mandatory, no?

If so, can't we just have something lile:
```python
if not isinstance(snakemake.input.idx, list):
    RAISE ERROR
index = path.splitext(snakemake.input.idx[0])[0]
```

votti commented:
I agree that this should be more specific, but would think that expecting *any* list may still be a bit vague.
Wouldn't the correct thing to check if at least all the files required are there?

I lack a bit the domain knowledge but would think, this should be all outputs of `bwa-mem2/index`: https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/bwa-mem2/index.html ?

So wouldnt the even better solution be to:
```
REQUIRED_IDX = {".0123", ".amb", ".ann", ".bwt.2bit.64", ".pac"}

idx_base, idx_ext = zip(path.splitext(index) for index in snakemake.input.idx)

missing_idx = REQUIRED_IDX - set(idx_ext)
if len(missing_idx) > 0:
    raise ValueError(f"Missing required indices: {missing_idx} declarad as input.idx")

index = idx_base[0]
```

fgvieira commented:
I am not super familiar with `bwa-mem2`, but I'd also say those are all outputs.
And your solution is definitely better!

votti commented:
Reading again through the discussion #522 I have 3 points:

1) I think the reference itself (`genome.fasta` in the [example](https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/bwa-mem2/mem.html)) would also be an implicitly required input.

Should this be added as well as part of this PR, eg as an explicit `ref` input?

2) I think the reference and the indices need to be in the same input folder and have matching naming - should this be validated as well?
-> i.e. validate that all indices have the `basename` of `{ref}` 

3) Even the [example](https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/bwa-mem2/mem.html) does not mention all the indices:
`multiext("genome.fasta", ".amb", ".ann", ".bwt.2bit.64", ".pac"),`

while the index wrapper additionally mentiones: `"{genome}.0123"

Is this an omission, or does BWA-MEM2 not require this?
At least I find this mentioned in the `BWA-MEM2` source code.

CC @lczech @dlaehnemann '

dlaehnemann commented:
Regarding your questions:

1. I don't think `samtools sort` actually needs the reference. It is not used in the command and the docs don't even mention a reference file for this command: https://www.htslib.org/doc/samtools-sort.html
   So I think, this was a mis-phrasing or misunderstanding in the original discussion. I would say, there's no need to add the reference itself on this wrapper.
2. With `1.` not necessary, this also shouldn't be necessary.
3. Yes, make it as complete as possible, and make it match what the `bwa-mem2/index` wrapper provides as output. :+1:  The mismatch you found most likely is an omission.

dlaehnemann commented:
By the way, many thanks for taking this up and for putting in the effort of fixing this properly!

votti commented:
Sorry for being a bit delayed in impelementing the "proper" solution: I got a strong flu last week and now battling catching up to dooming deadlines. After Wednesday I would have time to do this - if this is more urgent and someone else has time to finish this please go ahead!

votti commented:
I have now finished implementing this and also updated the tests and documentation.

What I found was that previously implemented approach of using `os.path.splitext` to identify the extension would actually not work correctly for an index such as `a.fasta.bwt.2bit.64` (where the extension is `.bwt.2bit.64').
I have now implemented a routine to identify the longest common prefix for all indices (i.e. `a.fasta`) and the extension from this.

fgvieira commented on bio/bwa-mem2/mem/wrapper.py:
```diff
@@ -27,11 +31,44 @@
-else:
-    index = path.splitext(snakemake.input.idx[0])[0]
+
+def get_common_prefix(strings: list[str]) -> str:
```
Isn't this function doing the same as [os.path.commonprefix](https://docs.python.org/3/library/os.path.html#os.path.commonprefix)?

votti replied:
Ah I did not know about this one. Thanks!
 Will change to it.

votti commented on bio/bwa-mem2/mem/wrapper.py:
```diff
@@ -27,11 +31,44 @@
+    raise ValueError(
+        "Could not determine common prefix of index files.\n"
+        "Please make sure that the index files are named correctly."
+    )
```
@fgvieira How do you think about this one - should I go for the custom exception class?

fgvieira replied:
I think it is a bit overkill, but up to you (this coderabitai thing is quite new).

dlaehnemann replied:
Johannes turned this `coderabbitai` application on. It's very chatty, but it does have some useful suggestions -- but also a lot of suggestions you can ignore. And you can chat back at it, telling it not to review certain things or why a certain suggestion is not useful. It then at least claims to learn from it... :sweat_smile:

fgvieira commented:
You also added a log file. Is that needed?

votti commented:
@log file and bam file: this were added by accident, I removed them from the corresponding commit.
Sorry!
